design.md

```diff
@@ -0,0 +1,138 @@
+# Kernel Design
+
+## Functional Requirements
+
+### gRPC Server
+
+The kernel runs a gRPC server that supports the following operations.
+
+#### Running Scripts
+
+- Operations:
+  - Run a script from its contents.
+  - Run a script from a Git File URI.
+- Dependency resolution:
+  - Scripts may import from other scripts in the project. These dependencies are resolved to the relevant files in the project.
+  - Scripts may depend on Bowler libraries specified by a Git File URI.
+    - By default, resolution uses GitFS to pull the script.
+    - If the URI resolves to a Bowler library that has been dev'd, then the dev'd version must be used instead of the version from the remote.
+  - If credentials are required to resolve a dependency, the kernel must try to load them from the local environment first. If that fails, the kernel must ask the IDE for authentication. The authentication received from the IDE must not be stored on disk. If no IDE is available, dependency resolution fails and an exception is thrown at the call site.
+- A script may be given arguments via a list of objects when it is invoked.
+- A script may return an object when it returns.
+- ! RFC: How are these objects serialized?
+
+#### GitFS
+
+- Each repo is uniquely identified by a Git Repo URI.
+- Each file in a repo is uniquely identified by a Git File URI.
+- Each Git URI includes a remote.
+- Operations:
+  - Clone or pull a repo.
+  - List files in a repo.
+  - Read the contents of a file in a repo.
+  - Run a script in a repo.
+    - May pass arguments as a list of objects.
+    - May receive a return value as an object.
+  - Load a robot from a robot config file in a repo.
+
+#### Hardware Management
+
+- Operations:
+  - Load a robot from its config file.
+    - May specify the contents of the config file.
+    - May specify a Git File URI.
+    - Must specify whether to connect to hardware or a simulator.
+      - In the hardware case, a connection method must be specified.
+      - In the simulator case, a simulator plugin must be specified.
+
+#### Simulator Plugins
+
+- Operations:
+  - List available plugins
+  - Add a plugin
+    - Adding a plugin with an ID that conflicts with an ID of an already-added plugin throws an error.
+    - ! RFC: What should the schema for adding a plugin be? The most natural method is by specifying the path to the Jar file, but we don't want to use file paths. We could support specifying a Git File URI. What else? I don't want to allow specifying content because that would involve sending potentially lots of megabytes over the RPC call. We could have a registry of simulator plugins similar to how we will have a registry of toolchains.
```

Comment:
There is nothing strictly wrong with "sending lots of megabytes over the RPC call", but I agree that there should be other options to avoid that. Would it be possible to specify a plugin that's hosted on maven, or a Groovy script as a plugin?

Reply:
I think that first, we should define what a simulator plugin is. I know that the data we need, in some form, is:
- A Jar
- Some metadata about what kernel versions the plugin is compatible with